espnet2/bin/asr_export_onnx.py

```diff
@@ -0,0 +1,184 @@
+import argparse
+
+import librosa
+import numpy as np
+import onnxruntime as rt
+import torch
+
+from espnet2.tasks.asr import ASRTask
+from espnet2.utils import config_argparse
+
+
+def export_encoder(encoder, feats_dim, out_filename="encoder.onnx", test_input=None):
+    """
+    Export an encoder module to ONNX.
+
+    Args:
+        encoder (nn.Module): Encoder module to export
+        feats_dim (int): Input feature dimension for the encoder
+        out_filename (str): Filename to save ONNX model to
+        test_input (tuple of (feats, feats_lens)):
+            An optional test input to verify the ONNX model.
+            If not given, test with the same dummy input used to export
+
+    Currently tested: conformer
+    """
+
+    dummy_feats = torch.randn(1, 200, feats_dim)
+    dummy_feats_lens = (
+        torch.ones(dummy_feats[:, :, 0].shape).sum(dim=-1).type(torch.long)
+    )
+    dummy_input = (dummy_feats, dummy_feats_lens)
+
+    torch.onnx.export(
+        encoder,
+        dummy_input,
+        out_filename,
+        verbose=False,
+        opset_version=15,
+        input_names=["feats", "feats_lens"],
+        output_names=["encoder_out", "encoder_out_lens"],
+        dynamic_axes={
+            "feats": {1: "feats_length"},
+            "encoder_out": {1: "enc_out_length"},
+        },
+    )
+
+    # test ONNX encoder and compare outputs with PyTorch encoder
+    if not test_input:
+        test_input = dummy_input
+
+    test_feats, test_feats_lens = test_input
+    enc_out, enc_out_lens, _ = encoder(test_feats, test_feats_lens)
+    enc_out = enc_out.detach().numpy()
+
+    options = rt.SessionOptions()
+    options.graph_optimization_level = rt.GraphOptimizationLevel.ORT_ENABLE_ALL
+    onnx_model = rt.InferenceSession(out_filename, options)
```

Comment:
We need to specify providers here with `onnxruntime>=1.9`.
We can add `providers=['CPUExecutionProvider']` to infer with CPU, or `'CUDAExecutionProvider` to use GPU.
Sample: https://onnxruntime.ai/docs/execution-providers/CUDA-ExecutionProvider.html#samples

Reply:
Can you explain more what specifying the provider does? If it's only for optimization purposes, I think it might not be necessary here, this call is mostly a sanity check to make sure that the export was successful. (in fact, I think we might   not even need the `graph_optimization_level`)

Reply:
@karthik19967829
Maybe your onnxruntime version is old. We need to specify providers when we use `onnxruntime>=1.9.0`, so would you test this script with the latest onnxruntime?
When I try to create InferenceSession with onnxruntime==1.11, I got this error.
```python
import onnxruntime as ort
m = ort.InferenceSession('./ctc.onnx')
```

> ValueError: This ORT build has ['CUDAExecutionProvider', 'CPUExecutionProvider'] enabled. Since ORT 1.9, you are required to explicitly set the providers parameter when instantiating InferenceSession. For example, onnxruntime.InferenceSession(..., providers=['CUDAExecutionProvider', 'CPUExecutionProvider'], ...)

Reply:
@kenzheng99 
Sorry, I mentioned the wrong person.